pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -2886,3 +3152,38 @@ func GetLoadBalancerSourceRanges(service *corev1.Service) (netsets.IPNet, error)
 	}
 	return ipnets, nil
 }
+
+// PreserveGopherError preserves the error details delivered with the response
+// that are explicitly omitted by dedicated error types.
+// error types from provider_client.go
+func PreserveGopherError(rawError error) error {
```

Comment:
@mandelsoft in case when gophercloud gets reauthenticated and it receives a 400 or 409 error on FIP allocation, you need to asser an error to `gophercloud.ErrErrorAfterReauthentication`, then extract the original error message:

```go
                if v, ok := err.(gophercloud.ErrErrorAfterReauthentication); ok {
                        err = v.ErrOriginal
                }
                // do further error assertion
```